### PR TITLE
Add withdraw and repay amount or max

### LINKED ISSUE #27: Create snippet with repay/withdraw max with only underlying amount as input

### PR DESCRIPTION
**Pull request by Jean-Grimal**
Fixes #27

### CODE REVIEW

File: src/blue/BlueSnippets.sol [hunk 1]
```diff
@@ -273,6 +273,32 @@ contract BlueSnippets {
...
+     * @return sharesWithdrawn The shares withdrawn in return for the assets.
+     */
+    function withdrawAmountOrAll(MarketParams memory marketParams, address user, uint256 amount)
+        external
+        returns (uint256 assetsWithdrawn, uint256 sharesWithdrawn)
+    {
+        uint256 shares;
+        address onBehalf = user;
+        address receiver = user;
+        uint256 totalSupplyAssets = morpho.expectedSupplyAssets(marketParams, user);
```
Comment (MerlinEgalite) [inline src/blue/BlueSnippets.sol:292]: we can accrue interest to save gas

File: src/blue/BlueSnippets.sol [hunk 2]
```diff
@@ -273,6 +273,32 @@ contract BlueSnippets {
...
+
+        if (amount >= totalSupplyAssets) {
+            shares = morpho.position(marketParams.id(), user).supplyShares;
+            (assetsWithdrawn, sharesWithdrawn) = morpho.withdraw(marketParams, 0, shares, onBehalf, receiver);
+        } else {
+            (assetsWithdrawn, sharesWithdrawn) = morpho.withdraw(marketParams, amount, shares, onBehalf, receiver);
```
Comment (MerlinEgalite) [inline src/blue/BlueSnippets.sol:298]:
```suggestion
            (assetsWithdrawn, sharesWithdrawn) = morpho.withdraw(marketParams, amount, 0, onBehalf, receiver);
```

File: src/blue/BlueSnippets.sol [hunk 3]
```diff
@@ -356,4 +382,29 @@ contract BlueSnippets {
...
+     * @return assetsRepaid The actual amount of assets repaid.
+     * @return sharesRepaid The shares repaid in return for the assets.
+     */
+    function repayAmountOrAll(MarketParams memory marketParams, address user, uint256 amount)
+        external
+        returns (uint256 assetsRepaid, uint256 sharesRepaid)
+    {
+        uint256 shares;
+        address onBehalf = user;
+        uint256 totalBorrowAssets = morpho.expectedTotalBorrowAssets(marketParams);
```
Comment (MerlinEgalite) [inline src/blue/BlueSnippets.sol:401]: same as above

File: src/blue/BlueSnippets.sol (hunk 1 above)
Comment (Rubilmax) [reply src/blue/BlueSnippets.sol:292]: with a comment indicating that if `view` is required (like in a `getWithdrawAmount` function), one can use `MorphoBalancesLib.expectedSupplyAssets`

File: src/blue/BlueSnippets.sol (hunk 3 above)
Comment (Rubilmax) [reply src/blue/BlueSnippets.sol:401]: https://github.com/morpho-org/morpho-blue-snippets/pull/30#discussion_r1444243665

File: test/forge/blue/TestBlueSnippets.sol [hunk 4]
```diff
@@ -222,6 +222,27 @@ contract TestIntegrationSnippets is BaseTest {
...
 
+    function testWithdrawAmountOrAll(uint256 amountSuplied, uint256 amountWithdrawn) public {
+        amountSuplied = bound(amountSuplied, 1, MAX_TEST_AMOUNT);
+        amountWithdrawn = bound(amountWithdrawn, 1, MAX_TEST_AMOUNT);
+        loanToken.setBalance(address(snippets), amountSuplied);
+
+        snippets.supply(marketParams, amountSuplied, address(snippets));
+        (uint256 assetsWithdrawn,) = snippets.withdrawAmountOrAll(marketParams, address(snippets), amountWithdrawn);
+
+        if (amountSuplied >= amountWithdrawn) {
```
Comment (Rubilmax) [inline test/forge/blue/TestBlueSnippets.sol:233]: We could bissect this test into 2 cases:
1. where `amountSupplied` is fuzzed to be >= `amountWithdrawn`
2. the other case

**Review by Rubilmax — APPROVED** [review]

**Review by MerlinEgalite — CHANGES_REQUESTED** [review]
to adapt since #31 has been merged

File: src/blue/BlueSnippets.sol [hunk 5]
```diff
@@ -48,10 +48,10 @@ contract BlueSnippets {
     // weird oracles / markets created
 
     /**
-     * @notice Calculates the supply APY (Annual Percentage Yield) for a given market.
-     * @param marketParams The parameters of the market.
-     * @param market The market for which the supply APY is being calculated.
-     * @return supplyRate The calculated supply APY.
+     * /// @notice Calculates the supply APY (Annual Percentage Yield) for a given market.
```
Comment (MerlinEgalite) [inline src/blue/BlueSnippets.sol:51]: weird haha

**Review by MerlinEgalite — CHANGES_REQUESTED** [review]

File: test/forge/blue/TestBlueSnippets.sol (hunk 4 above)
Comment (Jean-Grimal) [reply test/forge/blue/TestBlueSnippets.sol:271]: I'm fine with the current version but I'm okay to change if you want

Comment (Rubilmax) [reply test/forge/blue/TestBlueSnippets.sol:271]: The benefit is each side is fuzzed more, but it's fine for snippets to bissect this way

**Review by Rubilmax — APPROVED** [review]

**Review by MerlinEgalite — APPROVED** [review]

**Pull request merged**